### PR TITLE
Support for Asynchronous Tournaments

### PR DESCRIPTION
**Pull request by StanleyGoldman**
I'm using Coronate to conduct tournaments between friends similarly to how [lichess4545](https://www.lichess4545.com/) conducts a tournament.

Pairings are assigned and players are expected to play their games and report back within a timeframe.

Playing this way adds a few result states and I'd like to have them supported, probably behind a setting in the tournament as I'm sure it wouldn't apply to most use cases.

**Aborted** - Match never occured, no blame placed. No round points awarded. ELO should remain the same.
**WhiteAborted** - Match never occured, blame placed on White. Round point awarded to Black. ELO should remain the same.
**BlackAborted** - Match never occured, blame placed on Black. Round point awarded to White. ELO should remain the same.

### CODE REVIEW

**Comment by StanleyGoldman** [comment]
👋 @johnridesabike 

I'm slowly plodding my way though.
I'm unfamiliar with all the places I need to change to make this work.

If you got time to throw a pointer or two out.
Let me know.

**Comment by johnridesabike** [comment]
Thanks for the contribution! I don't think I have any objections to this kind of feature, at least not from glancing at it so far. 

Right now, I'd be fine keeping these new result types available by default without a setting. I've been conservative with adding new settings, since they increase the complexity of the overall app. I think having a few extra items on the dropdown menu won't be problem for general use.

Some points of interest:

The match results and player ratings are updated in this function: https://github.com/johnridesabike/coronate/blob/af065a85741de975e45fac4087dc34b6018ee22d/src/PageTournament/PageRound.res#L96

The UI for selecting match results is here: https://github.com/johnridesabike/coronate/blob/af065a85741de975e45fac4087dc34b6018ee22d/src/PageTournament/PageRound.res#L178

Match results are turned into scores with these functions: https://github.com/johnridesabike/coronate/blob/15cb838f925e494428aeaaf6115bd3eebfa25470/src/Data/Data_Scoring.res#L52 (There's a function for each player color.)

There are a few other places that will probably need to be updated too, but the sections linked above are the most important.

**Comment by StanleyGoldman** [comment]
I wholeheartedly agree with what you said about Settings.
My idea to hide this was going to be very cosmetic, I planned on just hiding these selections from the UI.

File: src/Data/Data_Scoring.res [hunk 1]
```diff
@@ -51,20 +51,20 @@ module Score = {
...
   let fromResultWhite = (x: Data_Match.Result.t) =>
     switch x {
-    | WhiteWon => One
-    | BlackWon => Zero
     | Draw => Half
     /* This loses data, so is a one-way trip. Use with prudence! */
-    | NotSet => Zero
+    | (WhiteWon | BlackAborted) => One
+    | (BlackWon | WhiteAborted) => Zero
+    | (NotSet | Aborted) => Zero
```
Comment (StanleyGoldman) [inline src/Data/Data_Scoring.res:58]: This feels concerning

File: src/Data/Data_Scoring.res [hunk 2]
```diff
@@ -51,20 +51,20 @@ module Score = {
...
   let fromResultBlack = (x: Data_Match.Result.t) =>
     switch x {
-    | WhiteWon => Zero
-    | BlackWon => One
     | Draw => Half
     /* This loses data, so is a one-way trip. Use with prudence! */
-    | NotSet => Zero
+    | (WhiteWon | BlackAborted) => Zero
+    | (BlackWon | WhiteAborted) => One
+    | (NotSet | Aborted) => Zero
```
Comment (StanleyGoldman) [inline src/Data/Data_Scoring.res:67]: Same here...

File: src/PageTournament/PageRound.res [hunk 3]
```diff
@@ -72,9 +72,14 @@ module MatchRow = {
...
       let lost = <Externals.VisuallyHidden> {React.string("Lost")} </Externals.VisuallyHidden>
+      let aborted = 
+        <span
+          ariaLabel="Draw" role="img" style={ReactDOMRe.Style.make(~filter="grayscale(60%)", ())}>
+          {React.string(`❌`)}
+        </span>
       switch m.result {
       | NotSet => <Externals.VisuallyHidden> {React.string("Not set")} </Externals.VisuallyHidden>
-      | Draw =>
+      | (Draw) =>
```
Comment (StanleyGoldman) [inline src/PageTournament/PageRound.res:82]:
```suggestion
      | Draw =>
```

File: src/PageTournament/PageRound.res [hunk 4]
```diff
@@ -104,6 +120,7 @@ module MatchRow = {
         | (_, None, _)
         | (_, _, None)
         | (NotSet, _, _) => (m.whiteOrigRating, m.blackOrigRating)
+        | (Aborted | WhiteAborted | BlackAborted, _, _) => (m.whiteOrigRating, m.blackOrigRating)
```
Comment (StanleyGoldman) [inline src/PageTournament/PageRound.res:123]:
```suggestion
        | (NotSet | Aborted | WhiteAborted | BlackAborted, _, _) => (m.whiteOrigRating, m.blackOrigRating)
```

File: src/PageTournament/PageRound.res [hunk 5]
```diff
@@ -103,7 +119,7 @@ module MatchRow = {
         let (whiteNewRating, blackNewRating) = switch (newResult, whiteOpt, blackOpt) {
         | (_, None, _)
         | (_, _, None)
-        | (NotSet, _, _) => (m.whiteOrigRating, m.blackOrigRating)
+        | (NotSet | Aborted | WhiteAborted | BlackAborted, _, _) => (m.whiteOrigRating, m.blackOrigRating)
```
Comment (StanleyGoldman) [inline src/PageTournament/PageRound.res:122]:
```suggestion
        | (Aborted | WhiteAborted | BlackAborted | NotSet, _, _) => (m.whiteOrigRating, m.blackOrigRating)
```

**Comment by StanleyGoldman** [comment]
@johnridesabike I'm testing this locally and everything seems to work out fine.

Otherwise thanks, this was fun.
I know F#, I have never programmed in OCaml or ReScript for that matter.

File: src/Data/Data_Scoring.res (hunk 1 above)
Comment (johnridesabike) [reply src/Data/Data_Scoring.res:58]: If you mean the comment, you can actually just remove that. There was a point where I was converting the score type into match results, and that comment was a note to myself to be careful with the "NotSet" variant. I don't think that happens anywhere now, and doing it wouldn't make sense anyway.

**Comment by johnridesabike** [comment]
Can you run `npm run test -- --updateSnapshot` and commit the changes? I think that should fix the failing test.

**Comment by johnridesabike** [comment]
The changes look good to me so far. One question though is how, or if, we want these new match results to affect and pairing and tiebreaking rules. For example, the auto-pair will avoid pairing players who played each other previously. Should that include players if their previous match was aborted? Also, should the "most black" tiebreaker count players who were assigned black in aborted matches? 

The answer to those may be "yes," and personally I'm fine with that. (The tiebreaking and pairing logic is already complex enough without adding new rules to it.) But either way it's worth considering.

**Comment by johnridesabike** [comment]
One more thing: please add a short entry to the changelog file. (You can just put it under the "latest" heading.)

Aside from that, I think it looks good!

**Comment by StanleyGoldman** [comment]
> One question though is how, or if, we want these new match results to affect the pairing and tiebreaking rules.

Some that happens to us often is a player in California is paired with a player in India. The match aborted because the players couldn't find the time in their schedules (considering time zones) to meet and play. Having the software avoid pairing those players again for the tournament makes sense because they likely wouldn't be able to meet up anyway.

I don't yet have a strong opinion about the tie breaking rule. I'll probably have an opinion after I need the tie break.

File: src/Data/Data_Scoring.res [hunk 6]
```diff
@@ -51,20 +51,20 @@ module Score = {
 
   let fromResultWhite = (x: Data_Match.Result.t) =>
     switch x {
-    | WhiteWon => One
-    | BlackWon => Zero
     | Draw => Half
     /* This loses data, so is a one-way trip. Use with prudence! */
```
Comment (StanleyGoldman) [inline src/Data/Data_Scoring.res:55]:
```suggestion
```

File: src/Data/Data_Scoring.res [hunk 7]
```diff
@@ -51,20 +51,20 @@ module Score = {
...
+    | (BlackWon | WhiteAborted) => Zero
+    | (NotSet | Aborted) => Zero
     }
 
   let fromResultBlack = (x: Data_Match.Result.t) =>
     switch x {
-    | WhiteWon => Zero
-    | BlackWon => One
     | Draw => Half
     /* This loses data, so is a one-way trip. Use with prudence! */
```
Comment (StanleyGoldman) [inline src/Data/Data_Scoring.res:64]:
```suggestion
```

File: src/Data/Data_Scoring.res (hunk 1 above)
Comment (StanleyGoldman) [reply src/Data/Data_Scoring.res]: Yep, I was talking about the comment.
Removed.

**Comment by johnridesabike** [comment]
Everything looks good to me. We can always tweak things later if it turns out we need to.

**Comment by StanleyGoldman** [comment]
Merge when ready

**Pull request merged**